### PR TITLE
BUG: Boxplots ignore xlabel and boxplot() doesn't support horizontal layout

### LINKED ISSUE #36918: BUG: Horizontal boxplots on subplots throws ValueError
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
import pandas as pd
import numpy as np

df1 = pd.DataFrame(np.random.randint(0,100,size=(100, 4)), columns=list('ABCD'))
df2 = pd.DataFrame(np.random.randint(0,100,size=(100, 4)), columns=list('ABCD'))

f, axs = plt.subplots(1, 2, figsize=(10,7), sharey=True)
df1.boxplot(ax=axs[0], showfliers=False, vert=False)
df2.boxplot(ax=axs[1], showfliers=False, vert=False)
```

#### Problem description

Above code snippet throws `ValueError: The number of FixedLocator locations (8), usually from a call to set_ticks, does not match the number of ticklabels (4).` due to changes introduced in matplotlib 3.3.0.

It looks like fixes were implemented as part of #35393 in `boxplot.py` for `vert=True` but not for `vert=False`. Performing the `len(ticks) != len(keys)` check for `ticks = ax.get_yticks()` and the associated keys list doubling snippet fixes the issue.

Not sure if that is the right fix, but works for my use case.

#### Expected Output
Two sets of horizontal boxplot charts

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 2a7d3326dee660824a8433ffd01065f8ac37f7d6
python           : 3.7.3.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-48-generic
Version          : #52-Ubuntu SMP Thu Sep 10 10:58:49 UTC 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.1.2
numpy            : 1.19.2
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 19.1.1
setuptools       : 41.0.1
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.5.2
html5lib         : None
pymysql          : None
psycopg2         : 2.8.6 (dt dec pq3 ext lo64)
jinja2           : 2.11.2
IPython          : 7.18.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.2
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : 1.5.2
sqlalchemy       : 1.3.19
tables           : None
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : None
numba            : None

</details>

### LINKED ISSUE #45463: BUG: Boxplots are ignoring xlabel
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd

df = pd.DataFrame(
    {'a': [2, 1, 4, 5, 1, 5], 'b': [4, 3, 5, 5, 2, 1]}
)
df.plot(kind='box', xlabel='x', ylabel='y')
df.plot(kind='box', xlabel='x', ylabel='y', vert=False)
df.boxplot(xlabel='x', ylabel='y')
```

### Issue Description

Results for the above snippet show that `xlabel` is ignored in both cases:

Screen Shot 2022-01-18 at 7 42 34 PM

Screen Shot 2022-01-18 at 7 43 39 PM

The final command raises an error:
<details>

```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/var/folders/2x/099whkgd0n99mrmnk__zzyr80000gp/T/ipykernel_17452/1525097035.py in <module>
----> 1 df.boxplot(xlabel='x', ylabel='y')

~/Desktop/pandas/pandas/plotting/_core.py in boxplot_frame(self, column, by, ax, fontsize, rot, grid, figsize, layout, return_type, backend, **kwargs)
    509 ):
    510     plot_backend = _get_plot_backend(backend)
--> 511     return plot_backend.boxplot_frame(
    512         self,
    513         column=column,

~/Desktop/pandas/pandas/plotting/_matplotlib/boxplot.py in boxplot_frame(self, column, by, ax, fontsize, rot, grid, figsize, layout, return_type, **kwds)
    423     import matplotlib.pyplot as plt
    424 
--> 425     ax = boxplot(
    426         self,
    427         column=column,

~/Desktop/pandas/pandas/plotting/_matplotlib/boxplot.py in boxplot(data, column, by, ax, fontsize, rot, grid, figsize, layout, return_type, **kwds)
    402             data = data[columns]
    403 
--> 404         result = plot_group(columns, data.values.T, ax)
    405         ax.grid(grid)
    406 

~/Desktop/pandas/pandas/plotting/_matplotlib/boxplot.py in plot_group(keys, values, ax)
    336         keys = [pprint_thing(x) for x in keys]
    337         values = [np.asarray(remove_na_arraylike(v), dtype=object) for v in values]
--> 338         bp = ax.boxplot(values, **kwds)
    339         if fontsize is not None:
    340             ax.tick_params(axis="both", labelsize=fontsize)

~/miniconda3/envs/test/lib/python3.9/site-packages/matplotlib/__init__.py in inner(ax, data, *args, **kwargs)
   1359     def inner(ax, *args, data=None, **kwargs):
   1360         if data is None:
-> 1361             return func(ax, *map(sanitize_sequence, args), **kwargs)
   1362 
   1363         bound = new_sig.bind(ax, *args, **kwargs)

TypeError: boxplot() got an unexpected keyword argument 'xlabel'
```

</details>

### Expected Behavior

Screen Shot 2022-01-18 at 7 44 56 PM

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : bc2a383c7615edadf2f1cd78541314ecb9b20857
python           : 3.9.6.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 20.6.0
Version          : Darwin Kernel Version 20.6.0: Tue Oct 12 18:33:42 PDT 2021; root:xnu-7195.141.8~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.4.0.dev0+1695.gbc2a383c76.dirty
numpy            : 1.20.3
pytz             : 2021.1
dateutil         : 2.8.2
pip              : 21.2.4
setuptools       : 52.0.0.post20210125
Cython           : 0.29.25
pytest           : 6.2.5
hypothesis       : 6.29.3
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.6.3
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.0.2
IPython          : 7.29.0
pandas_datareader: 0.10.0
bs4              : None
bottleneck       : 1.3.2
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.4.2
numba            : None
numexpr          : 2.7.3
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.7.1
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None

</details>

### PR DESCRIPTION
**Pull request by stefmolin**
- [X] closes #36918 and closes #45463
- [X] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/plotting/_matplotlib/boxplot.py [hunk 1]
```diff
@@ -237,20 +241,27 @@ def _grouped_plot_by_column(
...
-        naxes=naxes, sharex=True, sharey=True, figsize=figsize, ax=ax, layout=layout
+        naxes=naxes, sharex=kwargs.pop("sharex", True),
+        sharey=kwargs.pop("sharey", True), figsize=figsize, ax=ax, layout=layout
     )
 
     _axes = flatten_axes(axes)
 
+    # GH 45465: move the "by" label based on "vert"
+    xlabel, ylabel = kwargs.pop("xlabel", None), kwargs.pop("ylabel", None)
+    if kwargs.get("vert", 1):
```
Comment (mroeschke) [inline pandas/plotting/_matplotlib/boxplot.py:252]: Nit: Why 1 instead of True?

File: pandas/plotting/_matplotlib/boxplot.py [hunk 2]
```diff
@@ -332,18 +343,28 @@ def maybe_color_bp(bp, **kwds):
...
             ax.tick_params(axis="both", labelsize=fontsize)
-        if kwds.get("vert", 1):
-            ticks = ax.get_xticks()
-            if len(ticks) != len(keys):
-                i, remainder = divmod(len(ticks), len(keys))
-                assert remainder == 0, remainder
-                keys *= i
+
+        # GH 45465: x/y are flipped when "vert" changes
+        is_vertical = kwds.get("vert", 1)
```
Comment (mroeschke) [inline pandas/plotting/_matplotlib/boxplot.py:361]: Same here

File: pandas/tests/plotting/test_boxplot_method.py [hunk 3]
```diff
@@ -254,6 +256,53 @@ def test_specified_props_kwd(self, props, expected):
 
         assert result[expected][0].get_color() == "C1"
 
+    def test_xlabel_ylabel(self):
```
Comment (mroeschke) [inline pandas/tests/plotting/test_boxplot_method.py:259]: Could you split this into 3 distinct tests and use `pytest.mark.parameterize` over `vert`?

File: pandas/plotting/_matplotlib/boxplot.py (hunk 1 above)
Comment (stefmolin) [reply pandas/plotting/_matplotlib/boxplot.py:252]: That was how it was originally in the code before my edits. Happy to change it though 🙂

Comment (mroeschke) [reply pandas/plotting/_matplotlib/boxplot.py:252]: Gotcha. If you could, thanks!

Comment (stefmolin) [reply pandas/plotting/_matplotlib/boxplot.py]: Done.

File: pandas/plotting/_matplotlib/boxplot.py (hunk 2 above)
Comment (stefmolin) [reply pandas/plotting/_matplotlib/boxplot.py]: Done.

File: pandas/tests/plotting/test_boxplot_method.py (hunk 3 above)
Comment (stefmolin) [reply pandas/tests/plotting/test_boxplot_method.py]: Done.

**Review by mroeschke — APPROVED** [review]
LGTM

**Comment by mroeschke** [comment]
@stefmolin could you merge in the main branch once more? We had CI issues this past week

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @stefmolin

**Pull request merged**